jobs/sql/extractions/all_encounters.sql

```diff
@@ -0,0 +1,153 @@
+encounter_id, encounter_datetime, encounter_type, location_id, patient_id, visit_id, creator, entered_datetime)
+select encounter_id, encounter_datetime, encounter_type, location_id, patient_id, visit_id, creator, date_created 
+from encounter 
+where voided = 0
+;
+
+-- index
+CREATE INDEX temp_all_encounters_patientId ON temp_all_encounters(patient_id);
+CREATE INDEX temp_all_encounters_encounterId ON temp_all_encounters(encounter_id);
+CREATE INDEX temp_all_encounters_encounterType ON temp_all_encounters(encounter_type);
+CREATE INDEX temp_all_encounters_encounterLoc ON temp_all_encounters(location_id);
+CREATE INDEX temp_all_encountersCreator ON temp_all_encounters(creator);
+
+-- location name
+update temp_all_encounters t set encounter_location = location_name(t.location_id);  
+
+--  encounter type
+update temp_all_encounters t set encounter_type_name = encounter_type_name_from_id(t.encounter_type);
+
+-- user entered
+update temp_all_encounters t 
+inner join users u on t.creator  = u.user_id 
+set user_entered =  person_name(u.person_id) ;
+
+-- emr_id
+-- unique patients loaded to temp table before looking up emr_id and joining back in to the temp_all_encounters table
+drop temporary table if exists temp_emrids;
+create temporary table temp_emrids
+(patient_id		int(11),
+emr_id			varchar(50));
+
+insert into temp_emrids(patient_id)
+select DISTINCT patient_id
+from temp_all_encounters
+;
+create index temp_emrids_patient_id on temp_emrids(patient_id);
+
+update temp_emrids t set emr_id = patient_identifier(patient_id, 'ZL EMR ID');
+update temp_all_encounters t
+inner join temp_emrids te on te.patient_id = t.patient_id
+set t.emr_id = te.emr_id;
+
+-- next visit date
+-- all next appt date obs loaded to temp table before looking up for each encounter
+-- this avoids having to access the full obs table for each encounter
+DROP TABLE IF EXISTS temp_next_appt_obs;
+CREATE TEMPORARY TABLE temp_next_appt_obs
+select encounter_id, max(value_datetime) "next_appt_date"
+from obs where concept_id = @next_appt_date_concept_id
+and voided = 0
+group by encounter_id;
+
+create index temp_next_appt_obs_ei on temp_next_appt_obs(encounter_id);
+
+update temp_all_encounters te 
+inner join temp_next_appt_obs tvo on tvo.encounter_id = te.encounter_id
+set te.next_appt_date = tvo.next_appt_date;
+
+-- indexes
+drop temporary table if exists temp_all_encounters_index_asc;
+create temporary table temp_all_encounters_index_asc
+(
+	SELECT  
+           patient_id,
+           encounter_id,
+           encounter_datetime,
+		   index_asc
+FROM (SELECT  
+             @r:= IF(@u = patient_id, @r + 1,1) index_asc,
+             patient_id,
+             encounter_id,
+             encounter_datetime,
+			 @u:= patient_id
+            FROM temp_all_encounters,
+                    (SELECT @r:= 1) AS r,
+                    (SELECT @u:= 0) AS u
+            ORDER BY patient_id, encounter_id ASC
+            ) index_ascending );
+
```

Comment:
The above is fine, but I thought you implied yesterday that you realized this needs to be done in the warehouse after loading, maybe because encounters for a particular patient can come from more than one source DB

Reply:
Yeah, it's true.  I think I'd like to commit this as it is, see how it runs (performance etc...), and then assess how big an issue this is.  Moving this to the warehouse shouldn't be too hard but it's just something different than what we've done in the Haiti pipeline before.